apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/BodyCaptureImpl.java

```diff
@@ -0,0 +1,136 @@
+package co.elastic.apm.agent.impl.context;
+
+import co.elastic.apm.agent.objectpool.Resetter;
+import co.elastic.apm.agent.objectpool.impl.QueueBasedObjectPool;
+import co.elastic.apm.agent.tracer.configuration.WebConfiguration;
+import co.elastic.apm.agent.tracer.metadata.BodyCapture;
+import co.elastic.apm.agent.tracer.pooling.Allocator;
+import co.elastic.apm.agent.tracer.pooling.ObjectPool;
+import co.elastic.apm.agent.tracer.pooling.Recyclable;
+import org.jctools.queues.atomic.MpmcAtomicArrayQueue;
+
+import javax.annotation.Nullable;
+import java.nio.ByteBuffer;
+
+public class BodyCaptureImpl implements BodyCapture, Recyclable {
+    private static final ObjectPool<ByteBuffer> BYTE_BUFFER_POOL = QueueBasedObjectPool.of(new MpmcAtomicArrayQueue<ByteBuffer>(128), false,
+        new Allocator<ByteBuffer>() {
+            @Override
+            public ByteBuffer createInstance() {
+                return ByteBuffer.allocate(WebConfiguration.MAX_BODY_CAPTURE_BYTES);
+            }
+        },
+        new Resetter<ByteBuffer>() {
+            @Override
+            public void recycle(ByteBuffer object) {
+                object.clear();
+            }
+        });
+
+    private enum CaptureState {
+        NOT_ELIGIBLE,
+        ELIGIBLE,
+        STARTED
+    }
+
+    private CaptureState state;
+
+    private final StringBuilder charset;
+
+    /**
+     * The maximum number of bytes to capture, if the body is longer remaining bytes will be dropped.
+     */
+    private int numBytesToCapture;
+
+    @Nullable
+    private ByteBuffer bodyBuffer;
+
+    BodyCaptureImpl() {
+        charset = new StringBuilder();
+        resetState();
+    }
+
+    @Override
+    public void resetState() {
+        state = CaptureState.NOT_ELIGIBLE;
+        charset.setLength(0);
+        if (bodyBuffer != null) {
+            BYTE_BUFFER_POOL.recycle(bodyBuffer);
+        }
+    }
+
+    @Override
+    public void markEligibleForCapturing() {
+        if (state == CaptureState.NOT_ELIGIBLE) {
+            state = CaptureState.ELIGIBLE;
+        }
+    }
+
+    @Override
+    public boolean isEligibleForCapturing() {
+        return state != CaptureState.NOT_ELIGIBLE;
+    }
+
+    @Override
+    public boolean startCapture(@Nullable String charset, int numBytesToCapture) {
+        if (numBytesToCapture > WebConfiguration.MAX_BODY_CAPTURE_BYTES) {
+            throw new IllegalArgumentException("Capturing " + numBytesToCapture + " bytes is not supported");
+        }
+        if (state == CaptureState.ELIGIBLE) {
+            if (charset != null) {
```

Comment:
how can `charset` be null?

Reply:
`charset` is the parameter here, not the field (I'll rename the parameter to avoid confusion).

The charset may not always be available, e.g. if the `Content-Type` header was not present or did not contain a charset. In that case we'll later default to simply casting from `byte` to `char` (assuming the HTTP default charset).